### PR TITLE
Linearize bilinear terms of lower dual*primal in upper objective

### PR DESCRIPTION
**Pull request by NLaws**
## Summary
When the upper level objective contains bilinear products of lower dual and primal variables they can be linearized under certain conditions. The changes in this PR checks those conditions and linearizes the biinear terms when possible. The mathematical details can be found in:
http://www.optimization-online.org/DB_HTML/2021/08/8561.html

## Changes
- add an optional flag to `BilevelModel` called `linearize_bilinear_upper_terms`, which is a boolean that defaults to `false`.
- add automated test for the new capability using the existing `jump_conejo2016` test.
- address https://github.com/joaquimg/BilevelJuMP.jl/issues/158 by adding support for constructing `DualOf` vectors and DenseAxisArrays of constraint references

## Additional information
There are some additional, manual tests in a new file for the linearization process that I am keeping in the code for now. They will be useful to adapt and add to the automated tests as this new capability is expanded.
For example, currently the linearization process requires that the lower level model be provided by the user in its standard form. However, I believe that we could do the conversion of the lower level to standard form for users (and I have a start on it in the new functions). I will need help with that because I'm not sure how to handle the dualizing and all the maps if we convert the lower level to standard form before building the single level model. (I think that we will need a map from/to:
single level model to/from the dualized standard form lower model to/from the general form primal lower model).

### CODE REVIEW

**Comment by NLaws** [comment]
trying to close and reopen to kick off Actions tests

**Comment by NLaws** [comment]
@joaquimg I tried to get the tests running on actions by updating the testing framework to use test/Project.toml but it is still expecting the old format:
```
ERROR: Compat `Cbc` not listed in `deps` or `extras` section.
```
I can't find any info online on how to upgrade a package to the new Pkg testing framework. At first, Actions was failing because it could not add SCIP. Some guidance on how to get the test suite running would be greatly appreciated.

**Comment by joaquimg** [comment]
Awsome! I will take a look in the next days

**Comment by NLaws** [comment]
@joaquimg I am finding that for very large problems any loops over all variables or constraints takes a large amount of time. For example, the [build_bounds!](https://github.com/joaquimg/BilevelJuMP.jl/blob/059543614d1e8b6f59f285adeeb3ba8a43697242/src/jump.jl#L645) method loops over all variables and constraints when using `FortunyAmatMcCarlMode`. If there are ~20 lines of code that each take 0.001 seconds, then a problem with 1 million variables will take more than 5 hours to loop over all the variables. Is it possible to parallelize the `for` loops over variables and constraints? If you have suggestions for how to do that I can work on it in this PR.

**Comment by NLaws** [comment]
@joaquimg Can you please review this next chance you get? 
(The paper has been accepted to IEEE Transactions on Power Systems and will be published soon!)

**Comment by joaquimg** [comment]
I am trying to get to it. I had to stabilize some pieces of the code.
It is probably worth integrating master here so you can get xpress on CI.

**Comment by joaquimg** [comment]
Great news on the acceptance! They are usually very fast at posting a preview. Link here whenever there is something!

File: src/moi.jl [hunk 1]
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
...
+
+            if !(issubset(lower_con_types, standard_form_con_types))
+                linearize = false
+                @warn("The lower model must be in standard form to linearize bilinear terms, i.e. the constraint types must be a subset of $(standard_form_con_types). Skipping linearization process.")
+            else
+                A_N, bilinear_upper_dual_to_quad_term, bilinear_upper_dual_to_lower_primal, lower_primal_var_to_lower_con = 
+                    check_upper_objective_for_bilinear_linearization(upper, upper_to_lower_var_indices, upper_var_to_lower_ctr)
+
+                if isempty(A_N)
+                    @warn("No bilinear products of lower level dual and primal variables found in upper level objective. Skipping linearization process.")
```
Comment (joaquimg) [inline src/moi.jl:442]: this is only good for testing

File: src/moi.jl [hunk 2]
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
...
+                    AB_N, B, lower_obj_terms, lower_obj_type_handled = 
+                        get_lower_obj_coefs_of_upper_times_lower_primals(lower, lower_var_indices_of_upper_vars, A_N)
+                    if !lower_obj_type_handled
+                        @warn("Linearizing bilinear terms does not handle lower level objective type $(MOI.get(lower, MOI.ObjectiveFunctionType())). Skipping linearization process.")
+                        linearize = false
+                    end
+                end
+            end
+        else
+            @warn("Upper objective must be quadratic and there must lower dual variables in the upper model to linearize bilinear terms. Skipping linearization process.")
```
Comment (joaquimg) [inline src/moi.jl:454]: also only for tests right?

File: src/moi.jl [hunk 3]
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
...
+                        lower_obj_terms,
+                        lower_to_m_idxmap,
+                        lower_primal_dual_map,
+                        lower_dual_idxmap
+                    )
+                    if isnothing(linearizations)
+                        @warn("Unable to linearize bilinear terms due to underdertermined system of equations. Skipping linearization process.")
+                    end
+                else
+                    @warn("Required conditions for linearization not met. Skipping linearization process.")
```
Comment (joaquimg) [inline src/moi.jl:515]: can you say which? referencing the paper seems fine.

File: src/moi.jl (hunk 1 above)
Comment (joaquimg) [reply src/moi.jl:442]: these ones that are only for tests should be `@debug`

File: src/moi.jl [hunk 4]
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
...
+                end
+
+            end
+
+            # LINEARIZATION
+            # TODO switch signs with MAX sense?
```
Comment (joaquimg) [inline src/moi.jl:564]: maybe for now add an assert that we have a min sense

File: src/moi.jl [hunk 5]
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
...
+            if !(isnothing(linearizations))
+                # set m's objective by replacing quadratic terms with linearizations
+                mobj = deepcopy(m_objective)
+                quadratic_terms = mobj.quadratic_terms  # TODO keep quadratic terms that have not been linearized
+                c = mobj.constant
+                affine_terms = mobj.affine_terms
+                cvb = collect(values(bilinear_upper_quad_term_to_m_quad_term))
+                new_objective = deepcopy(m_objective)
+                if Set(cvb) == Set(quadratic_terms)
+                    @info("Replacing bilinear lower dual * lower primal terms in upper objective with linear terms.")
```
Comment (joaquimg) [inline src/moi.jl:575]: `@debug`

File: src/bilinear_linearization.jl [hunk 6]
```diff
@@ -0,0 +1,811 @@
+#=
```
Comment (joaquimg) [inline src/bilinear_linearization.jl:1]: we really need small test cases for all corner cases.

**Review by joaquimg — CHANGES_REQUESTED** [review]
This is a great contribution and I am keen on merging when (minimally) ready. Thank you so much for giving it a try.

1 - I would be fine merging a version that throws errors/warnings if strict conditions is not satisfied.
However, we really need a set of small tests covering all corner cases, so that I can improve performance and organization later on without silently breaking things.

2 - if you can put renames and vector forms DualOf in a separate PR we can merge them much earlier (which would already be good contributions)

3 - we need to rebase this by integrating current master where Ci is smoother.

4 - I will think more about the conversion to standard form. My current take is: if linearization is active:
before doing anything in build_bilevel
a - check if LL is LP and there are products that have a chance to be linearized (this is basically checking ocntraint type and loop in the upper objective) if all goo continue to b, otherwise skip the linearization.
b - create an MOI model for the standard lower level (keep important maps from the original problem to this (this can be done with a copy_to moving data from the original data to a MOI model that only supported standard form so that MOI bridges will do all conversions for you) - keep maps from this standardized model to the original LL
c - extract matrices using MatrixOfConstraints using another copy_to (we can follow DiffOpt.jl example)
d - check all other conditions from matrices in step c, it fails forget linearization and proceeded as usual, if good continue to e
e - use the model from b as if you were using the original LL - we will have to be careful with maps so that solution can be recovered, but this is doable.

5 - As for build bounds, there are probably lots of type instability. Can you create an issue with a MWE that takes like 5 seconds (that should be zero). Smallest problem that you can see some meaningful slowness...

6 - I tried to respond to all your previous comments in this single shot, please let me know if I missed any.

And again, awesome work, lets try to merge this asap.

**Comment by NLaws** [comment]
Thank you @joaquimg! I will respond to your comments that I have answers for now; and I will work on MWEs/more tests soon.

File: test/bilinear_linearization.jl [hunk 7]
```diff
@@ -0,0 +1,339 @@
+
+using BilevelJuMP
+# using Ipopt
+# using Cbc
+# using CPLEX
+using Gurobi
```
[the remark this answers is not in the document]
Comment (NLaws) [reply test/bilinear_linearization.jl:6]: I do have access to Xpress and I have merged the latest master. The Xpress/CI tests are passing for this PR (but the codecov tests are not). What do you recommend to address the codecov tests? (I am unfamiliar with them, but it looks like I need more doc strings and/or tests?)

File: test/bilinear_linearization.jl [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    number of non-zero qp constraint terms = 0
+
+So I assumed that it is not modeling the lower mu SOS1 constraints, ran Gurobi NLP again, and get
+the result that matches BilevelJuMP solution (UL cost 0, LL cost 3)
+
+So it seems that BilevelJuMP (actually it _is_ Dualization.jl dual_problem.primal_dual_map in BilevelJuMP moi.jl)
+ is not including the complementary constraints for lower bounds of zero 
+(but is including the dual variables for lower bounds of zero in the dual constraints, which results in the odd solution of UL cost 0, LL cost 3)
+
+Bug in v 0.3.5 of Dualization.jl see https://github.com/jump-dev/Dualization.jl/issues/123 and address by checking out master branch of BilevelJuMP (Done)
```
[the remark this answers is not in the document]
Comment (NLaws) [reply test/bilinear_linearization.jl:180]: done. I will remove these notes.

File: src/moi.jl (hunk 1 above)
Comment (NLaws) [reply src/moi.jl:442]: I actually meant all @warn's for users. In the case that a user provides `linearize_bilinear_upper_terms=true` but the model does not contain valid bilinear products, then this warning will notify the user that their model is not  a candidate for linearization.

File: src/moi.jl (hunk 2 above)
Comment (NLaws) [reply src/moi.jl:454]: Meant for users that provide `linearize_bilinear_upper_terms=true` but do not have a model that is appropriate for linearization.

File: src/moi.jl (hunk 3 above)
Comment (NLaws) [reply src/moi.jl:515]: Yes the @debug statements tell users which conditions were not meant. Would you prefer that the unmet conditions are collected into this @warn statement?

File: src/moi.jl (hunk 5 above)
Comment (NLaws) [reply src/moi.jl]: done in bc4539998c421fec88da7461efe8ebcdbdb6d0b9

File: src/moi.jl (hunk 4 above)
Comment (NLaws) [reply src/moi.jl]: done in 3f96626ceef9713d84f5585e3a2d9621d843d8bb

File: src/moi.jl [hunk 9]
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
         pass_names(m, lower_dual, lower_dual_idxmap)
     end
 
+    #=
+        Check if Upper bilinear terms of the form Upper primal * Lower dual exist, and can be linearized
+    =#
+    # TODO implement for bilinear products in constraints (only checking upper objective so far)
+
+    if linearize_bilinear_upper_terms
```
[the remark this answers is not in the document]
Comment (NLaws) [reply src/moi.jl:421]: done in 5ba028ee13fac95bce8350b5656b44fd625eb912

File: src/moi.jl (hunk 3 above)
Comment (joaquimg) [reply src/moi.jl]: in this case I meant, lets add more details about what condition is failing to be met, keep as warning.

File: test/bilinear_linearization.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (joaquimg) [reply test/bilinear_linearization.jl]: for codecov we basically need more small tests testing different bits of the overall code. It will tell you which pieces of code are never reached, so that you can add testes that reach those lines.
Focus on your added functions.

[the remark this answers is not in the document]
Comment (joaquimg) [reply test/bilinear_linearization.jl]: dont worry about docstrings now

File: src/moi.jl (hunk 1 above)
Comment (joaquimg) [reply src/moi.jl]: I mean, if linearization is active but there is nothing to do, just use `@debug`, but if it failed to linearized due to some condition not being met, then use `@warn`

File: src/moi.jl (hunk 2 above)
Comment (joaquimg) [reply src/moi.jl]: I mean, if linearization is active but there is nothing to do, just use @debug, but if it failed to linearized due to some condition not being met, then use @warn

Comment (NLaws) [reply src/moi.jl]: Got it, addressed in b02f8bc

File: src/jump_constraints.jl [hunk 10]
```diff
@@ -169,16 +169,60 @@ end
 struct DualOf
     ci::BilevelConstraintRef
 end
+
+function DualOf(cis::AbstractVector{<:BilevelConstraintRef})
+    return [DualOf(ci) for ci in cis]
+end
+
+function DualOf(cis::JuMP.Containers.DenseAxisArray{<:BilevelJuMP.BilevelConstraintRef})
```
[the remark this answers is not in the document]
Comment (NLaws) [reply src/jump_constraints.jl:177]: > 1 - Maybe a single method defined for `AbstractArray` calling `similar` to initialize and `eachindex` to fill would work for all 3 main cases: `Array`, `DenseAxisArray`, `SparseAxisArray`

I'm not sure what you mean. Can you please implement or provide an example?

File: src/moi.jl (hunk 3 above)
Comment (NLaws) [reply src/moi.jl]: addressed in b02f8bc

File: src/moi.jl (hunk 1 above)
Comment (NLaws) [reply src/moi.jl]: got it. change to debug in 0d7c8a0

**Comment by joaquimg** [comment]
That is great!

**Comment by NLaws** [comment]
@joaquimg how does this codecov thing work? I added an explicit test for one of my new functions https://github.com/joaquimg/BilevelJuMP.jl/pull/157/commits/d1e893e828bee3719a90770c85afd5f1edb9b2e1 but the [codecov still shows that this code is untested]( https://app.codecov.io/gh/joaquimg/BilevelJuMP.jl/compare/157/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr%20comments&utm_term=Joaquim%20Dias%20Garcia#D4L761)?

**Comment by joaquimg** [comment]
codecov will literally check which lines were touched during tests.
it some rare cases, it might count something wrong due inlining.

**Comment by joaquimg** [comment]
Now `master` branch is in JuMP 1.0!
So you can use update versions of anything if you need.

File: src/jump_constraints.jl (hunk 10 above)
[the remark this answers is not in the document]
Comment (NLaws) [reply src/jump_constraints.jl:199]: I used your new unit test for `DualOf` a vector constraint

File: test/bilinear_linearization.jl [hunk 11]
```diff
@@ -0,0 +1,339 @@
+
```
[the remark this answers is not in the document]
Comment (NLaws) [reply test/bilinear_linearization.jl:5]: done

File: test/bilinear_linearization.jl [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+# complementary slackness
+@constraint(model, [y[1]-300, mu_1_up] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [y[2]-150, mu_2_up] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [y[3]-100, mu_3_up] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [y[1], mu_1_lo] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [y[2], mu_2_lo] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [y[3], mu_3_lo] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [s, mu_s_lo] in MOI.SOS1([1.0, 2.0]))
+# seems like maybe the mu variables are not constrained to be positive in BilevelJuMP ? have to add these constraints manually?
```
[the remark this answers is not in the document]
Comment (NLaws) [reply test/bilinear_linearization.jl]: I don't think this was the issue (hence the ?, it was more a note to self).

File: src/bilinear_linearization.jl (hunk 6 above)
Comment (NLaws) [reply src/bilinear_linearization.jl:1]: I am adding more tests and trying to meet the codecov requirements

File: src/jump_constraints.jl [hunk 13]
```diff
@@ -191,29 +191,60 @@ end
...
-        "Either do:\n" *
-        "@variable(Upper(m), my_variable[t=1:N], " *
-        "DualOf(my_constraint_vector[t]))\n" *
-        "Or use anonynous variables:\n" *
-        "@variable(Upper(m), variable_type = DualOf(my_constraint_vector[t]))"
-    )
+
+function DualOf(cis::AbstractVector{<:BilevelConstraintRef})
+    return [DualOf(ci) for ci in cis]
+end
```
[the remark this answers is not in the document]
Comment (joaquimg) [reply src/jump_constraints.jl:197]: The error message was detailing work arounds that are more "JuMP compliant"

[the remark this answers is not in the document]
Comment (NLaws) [reply src/jump_constraints.jl:197]: Sorry, I'm not understanding. What does the form of the right hand side of a constraint have to do with the `DualOf` function?

[the remark this answers is not in the document]
Comment (NLaws) [reply src/jump_constraints.jl:197]: Are you saying that [this](https://github.com/NLaws/CoDER_examples/blob/4e9667b5fce2bdf9cad0664df3207fab5526bc91/planning_bileveljump.jl#L63) is wrong?
```julia
# LL load balance with PV (lambda)
@constraint(Lower(model), loadbalance[n in LLnodes_withPV, t in 1:T], 
    -ye[n,t] + yi[n,t] + ypvprod[n,t] - LDFinputs.Pload[n][t] == 0
);

# duals of LL load balances
@variable(Upper(model), 0 <= lambda <= M, DualOf(loadbalance));
```

**Comment by NLaws** [comment]
@joaquimg any idea why non-linear tests would start failing?
```julia
Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2548
  Expression: ≈(objective_value(model), 12, atol = 0.1)
   Evaluated: 0.0 ≈ 12 (atol=0.1)

Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2549
  Expression: ≈(value(x), 6, atol = 0.001)
   Evaluated: 1.9596051811680095 ≈ 6 (atol=0.001)

Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2551
  Expression: ≈(value(y[1]) - value(y[2]), 2, atol = 0.001)
   Evaluated: 1.5098463839385659 ≈ 2 (atol=0.001)
```

**Comment by NLaws** [comment]
> @joaquimg any idea why non-linear tests would start failing?

The failures seem random. Now they are passing (or maybe they are not always running?)

**Comment by NLaws** [comment]
@joaquimg I'm not sure what else I can do about the codecov. From what I can tell it is now dinging me for coverage reductions for lines that I did not add. For example, it says that there is a -57% coverage reduction for intervals.jl and a -31% reduction in moi.jl.

**Comment by joaquimg** [comment]
> @joaquimg any idea why non-linear tests would start failing?
> 
> ```julia
> 
> Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2548
> 
>   Expression: ≈(objective_value(model), 12, atol = 0.1)
> 
>    Evaluated: 0.0 ≈ 12 (atol=0.1)
> 
> 
> 
> Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2549
> 
>   Expression: ≈(value(x), 6, atol = 0.001)
> 
>    Evaluated: 1.9596051811680095 ≈ 6 (atol=0.001)
> 
> 
> 
> Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2551
> 
>   Expression: ≈(value(y[1]) - value(y[2]), 2, atol = 0.001)
> 
>    Evaluated: 1.5098463839385659 ≈ 2 (atol=0.001)
> 
> ```

Ipopt can converge to bad points.
New versions, change or orders of things in the models, different OS, might lead to that.

**Comment by joaquimg** [comment]
> @joaquimg I'm not sure what else I can do about the codecov. From what I can tell it is now dinging me for coverage reductions for lines that I did not add. For example, it says that there is a -57% coverage reduction for intervals.jl and a -31% reduction in moi.jl.

I will look carefully through coverage later. Focus your coverage work in you new code. Try keep adding tests that result in different paths your code can go. That will be enough for me.
I can do the polishing.

**Comment by NLaws** [comment]
@joaquimg I think that we are good with test coverage of my contributions: ~400 out of ~1600 of the new lines are tests and src/bilinear_linearization.jl has ~95% coverage. Please let me know if there is anything specific that you would like me to address.

**Comment by joaquimg** [comment]
Awesome. I will do another pass.

**Comment by NLaws** [comment]
@joaquimg are you going to merge this PR? If not I will close it.

**Comment by joaquimg** [comment]
I love this PR and I intend to move forward with this, yes.
Sorry for the delay.
My biggest issue is that it does not apply in all cases, which is totally fine. However, we need to keep this clear for the users to minimize confusions. We need clear error messages so the user understands that assumptions are being violated.
That might be already done, but I have not had time to dive again into it.
Will try this week.

**Comment by joaquimg** [comment]
This is getting very nice.
Let me know whenever it is ready for another pass!

**Comment by NLaws** [comment]
@joaquimg ^^

**Comment by NLaws** [comment]
@joaquimg ^^
